app/components/Account/CollateralTable.jsx

```diff
@@ -0,0 +1,255 @@
+import Translate from "react-translate-component";
+import counterpart from "counterpart";
+import {List} from "immutable";
+import TranslateWithLinks from "../Utility/TranslateWithLinks";
+import Immutable from "immutable";
+
+const alignRight = {textAlign: "right"};
+const alignLeft = {textAlign: "left"};
+
+/**
+ * Renders whole margin list including call orders and placeholders
+ */
+class ListGenerator extends React.Component {
+    static propTypes = {
+        callOrders: ChainTypes.ChainObjectsList
+    };
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            assetsPropsCount: 0,
+            callOrdersCount: 0
+        };
+    }
+
+    static getDerivedStateFromProps(nextProps, prevState) {
+        // if we got new bitasset info
+        // or call orders changed
+        // then recalculate "cache" of maring items
+
+        const callOrdersJson = JSON.stringify(nextProps.callOrders);
+
+        if (
+            prevState == null ||
+            prevState.assetsPropsCount != nextProps.bitAssets.length ||
+            prevState.ordersJson != callOrdersJson
+        ) {
+            let assets = new Array();
+            // construct map of bitassets
+            let assetsMap = {};
+            // index to track asset info
+            let index = 0;
+
+            nextProps.bitAssets.forEach(o => {
+                assetsMap[o.get("id")] = index++;
+                assets.push({
+                    asset: o,
+                    has_margin_order: false
+                });
+            });
+
+            if (nextProps.callOrders.length > 0) {
+                // add to iterated bitasets items that wasn't listed by default (component's callOrders prop)
+                nextProps.callOrders.forEach(o => {
+                    let assetId = o.getIn(["call_price", "quote", "asset_id"]);
+
+                    if (assetsMap[assetId] == null) {
+                        let newAssetInfo = ChainStore.getObject(assetId);
+                        if (typeof newAssetInfo != "undefined") {
+                            assetsMap[assetId] = index++;
+                            assets.push({
+                                asset: newAssetInfo,
+                                has_margin_order: true,
+                                order: o
+                            });
+                        }
+                    } else {
+                        // mark as margin order
+                        assets[assetsMap[assetId]].has_margin_order = true;
+                        assets[assetsMap[assetId]].order = o;
+                    }
+                });
+            }
+
+            return {
+                assetsPropsCount: nextProps.bitAssets.length,
+                assets: assets,
+                ordersJson: callOrdersJson
+            };
+        }
+
+        return prevState;
+    }
+
+    render() {
+        let {account, callOrders} = this.props;
+        callOrders = callOrders.filter(o => !!o);
```

Comment:
Does this really happen, I mean that `o` does not exist?
In `getDerivedStateFromProps` this check is not included. 

In any case: 
- Is this used anywhere further after filtering?
- If so, would this not be more suitable to be filtered when receiving props, or when actually using it to render?

Reply:
yup, sometimes (when ws client just connect to api) we got some "undefined" results when using.
yes, getDerivedStateFromProps should use this too

Reply:
and yes, `callOrders` not used in render method anymore -  handle those data in `getDerivedStateFromProps`, so this line is going to be removed

Reply:
Where is callOrders used again in the rendering? Can't find it in MarginPosition